### PR TITLE
26377 merge actions

### PR DESCRIPTION
**Pull request by nylar**
Merges the old action branch into master, 100% test coverage, tested in dev on the exchange (although I am skeptical :stuck_out_tongue: )

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -58,12 +59,12 @@ func NewClient(host string, interval int) (cl *Client, err error) {
...
 		hostURL:    host,
 		http:       &http.Client{},
-		logger:     log.New(os.Stderr, "", log.LstdFlags), // Stderr logging
+		logger:     log.New(os.Stderr, "", log.Ldate|log.Ltime|log.Lshortfile),
 		interval:   intDur,
 		input:      make(chan MetricWithValue),
 		stop:       make(chan bool, 1),
 		stopped:    make(chan bool, 1),
-		metrics:    make(map[Metric]int, 1000),
+		metrics:    make(map[Metric]Value),
```
Comment (manavo) [inline client.go:67]: This changed from a buffered channel to an unbuffered one. Intentional or accidental?

File: client.go [hunk 2]
```diff
@@ -87,50 +88,92 @@ func newBufferPool() *sync.Pool {
...
+		v.Sum = &Sum{
+			Value: value,
+		}
+	case "avg":
+		v.Avg = &Average{}
+
+		var avgResult int
+		var newCount int
+
+		if val, ok := c.metrics[m]; ok {
```
Comment (manavo) [inline client.go:126]: Not sold on this. Access to `c.metrics` should be dependant on the mutex, right? Maybe these calculations need to be happening in the `handleMetricWithValue` function? Don't think it is thread safe this way!

Comment (nylar) [inline client.go:126]: Yup definitely, wrapping the send contents in a mutex lock. Also, a race condition in the flush function checking the length of the map, moved this check after the mutex lock.

Edit: It was in the original actions branch but was lost in transit, will add it back in.

Comment (matzhouse) [inline client.go:126]: Yeah - it's a valid point. c always needs to be protected. 

I'm wondering if all the functions that use send() (Count, Timer, AverageTimer) should call it in a goroutine - make it non blocking at the first call.

Those calc functions are a little mixed up to me.. sum sets a value but doesn't do anything to c, but then avg goes and works out the value and updates c.. not sure we need to chan if we just go and set the values on c - and if all the inputs are in goroutines this is all non blocking anyway..

Comment (matzhouse) [inline client.go:126]: I merged this thinking i'd reviewed it on friday and it was ok - but I suspect i'm lying to myself - apologies :)

**Pull request merged**